### ISSUE TITLE
有关ppo reward被mask掉的问题

### ISSUE THREAD

**Issue by Ervinoreo** [wyf3/llm_related#37]
up主你好，感谢你的分享。有个小小疑惑，假如ends=action_mask.sum(1) + 1, 不会出现代码下面那个例子的问题吗？

```python
def compute_rewards(kl, r, action_mask, kl_ctl, clip_reward_value):

        kl_divergence_estimate = -kl_ctl * kl
        rewards = kl_divergence_estimate

        ends = action_mask.sum(1) + 1
        
        if not isinstance(clip_reward_value, torch.Tensor):
            clip_reward_value = torch.tensor(clip_reward_value).to(r.device)
    
        reward_clip = torch.clamp(r, -clip_reward_value,
                                  clip_reward_value)
        batch_size = r.size(0)
        for j in range(batch_size):
            rewards[j, :ends[j]][-1] += reward_clip[j, 0]

        return rewards


action_mask = [1, 1, 1, 1, 0, 0]  # 4 valid tokens
ends = 5
rewards[j, :5][-1]  # = rewards[j, 4] ← EOS position!
# But action_mask[4] = 0, so this reward gets masked out in loss! ❌

**Comment by wyf3** [comment]
感谢指出，已经调整
